README
Tutu - A skimpy outfit for Dancer 

Tutu is a flat-file CMS inspired by Blosxom.  Pages are
created as textfiles, typically formatted with Markdown.
Tutu is based on the Dancer web framework.
 
Tutu is configured as a website to display pages and to
distribute media files for training course. It requires
login, authenticated users may upload and download files.

INSTALLATION

To install this module, you may run the following commands:

	perl Makefile.PL
	make
	make test
	make install

CUSTOMIZING THE DEFAULT WEBAPP TEMPLATE

	Edit MY_WEBSITE_INIT to suit your planned site.
	Run this script:

    ./customize.pl

	Test your app by starting the local webserver and
	pointing your browser to http://localhost:3000

	cd $WEBSITE_ROOT

	./dance.psgi

    Move the $WEBSITE_ROOT and $WEBSITE_DOMAIN directories
    to the appropriate locations, if desired.

COPYRIGHT AND LICENCE

Copyright (C) 2014 Joel Roth

This program is free software; you can redistribute it and/or modify it
under the same terms as Perl itself.


Changes
Revision history for Tutu

0.01    May 24, 2014

        Initial verison, stripping site-specifics
		renaming modules to Tutu and Tutu::Utils


Makefile.PL
use strict;
use warnings;
use ExtUtils::MakeMaker;

WriteMakefile(
    NAME                => 'Tutu',
    AUTHOR              => 'Joel Roth <joelz@pobox.com>',
    VERSION_FROM        => 'lib/Tutu.pm',
    ABSTRACT_FROM       => 'lib/Tutu.pm',
    PL_FILES            => {},
    PREREQ_PM => {
        'Test::More'				=> 0,
        'Dancer' 					=> 0,
		'Cwd' 						=> 0,
		'Crypt::PasswdMD5' 			=> 0,
		'Data::Dumper::Concise' 	=> 0,
		'Dancer ' 					=> 0,
		'Dancer::Error' 			=> 0,
		'Dancer::Session' 			=> 0,
		'Email::Stuffer' 			=> 0,
		'File::Basename' 			=> 0,
		'File::Find::Rule' 			=> 0,
		'File::Slurp' 				=> 0,
		'Modern::Perl' 				=> 0,
		'Text::Markdown' 			=> 0,
    },
	EXE_FILES => [ 'bin/tutu-index', 'bin/tutu-pass' ],
    dist                => { COMPRESS => 'gzip -9f', SUFFIX => 'gz', },
    clean               => { FILES => 'Tutu-*' },
);


customize.pl
#!/usr/bin/env perl
use Modern::Perl;
use Data::Dumper::Concise;

my @customize = split "\n",qx(cat ./MY_WEBSITE_INIT);
my %replace;
my $rep;
for (@customize)
{
	chomp;
	my ($var, $replacement) = split " ", $_, 2;
	$var =~ s/\s+$//;		
	$replacement =~ s/\s+$//;		
	$replacement =~ s{([\@\$])}{\\$1}g;

	$replace{$var} = $replacement;
	
	$var =~ s{\$}{\\\$};
	$rep .= "s{$var}{$replacement}; ";
}

my $new_domain_dir = $replace{'maxfankel.com'};
my $copy_domain_dir_cmd = "cp -a example.com $new_domain_dir";
say "
copying domain directory: 
    $copy_domain_dir_cmd
";
-d $new_domain_dir 
	and say ("directory $new_domain_dir already exists, fix this and try again."),
	exit
or system $copy_domain_dir_cmd;

my ($new_webapp_dir) = $replace{'/home/maxfankel/space'} =~ m{/?([^/]+)/?$};

my $copy_webapp_dir_cmd = "cp -a webapp $new_webapp_dir";
say "
copying webapp directory: 
    $copy_webapp_dir_cmd
";
-d $new_webapp_dir 
	and say ("directory $new_webapp_dir already exists, fix this and try again."),
	exit
or system $copy_webapp_dir_cmd;

my $grep = q(grep -lrP '\$[A-Z]'  $new_webapp_dir);
my $targets = join " ",map{chomp; $_} qx($grep);
say "
Searching for files using shell command: 

    $grep
";
say "Found targets: 

    $targets";

# escape sigils


#say Dumper \%replace;
$rep =~ s{'}{&#39;}g;
my $command = qq(perl -p -i.bak -E '$rep' $targets);
say;
say $command;
`$command`;


lib/Tutu/Utils.pm
package Tutu::Utils;
use Modern::Perl;
#use HTML::GenerateUtil qw(escape_uri);
use autodie;
use Data::Dumper::Concise;
use Crypt::PasswdMD5;
use Cwd;
use Email::Stuffer;
use File::Slurp;
require Exporter;
our @ISA = qw(Exporter);
our @EXPORT_OK = qw( index_directories 
					generate_passwd 
					generate_user_list
					find_user 
					send_user_password 
					remote_reminder 
					send_out_mail);

use vars qw( $website_name $website_url $webmaster_name $webmaster_mail_address );
our $sender = $webmaster_mail_address;

sub index_directories { 
	my $root = shift // getcwd;
	my $fh;
	my $public = "$root/public";
	my $pages  = "$root/pages";
	#$ENV{LANG} = "C";  # dates in non-English locales not supported
	opendir my ($dirh), $public;
	my %exclude = map { $_ => 1 } qw( . .. css images );
	my @top = grep{ ! $exclude{$_} and -d "$public/$_" } readdir $dirh; 
	#say "Top-level: @top";
	my $i++;
	map{
			dir_list(
					source => $public,
					target => $pages,
					current => $_,
					path => [],
			);
	} @top;
}

sub dir_list {
        my %args = @_;
        #print Dumper \%args;
        my @path = $args{path} ? @{ $args{path} } : ();
        my $source = $args{source};
        my $current = $args{current};
        my $target  = $args{target};
        my $output_file = join '/',$target,@path,$current.'.md';
        my $full = join '/',$source,@path,$current;
        chdir $full;

        opendir my($dir), $full;
        # sort order: directories, then files, ignore case
        my @fobj = sort { my $oa = -d $a ? "0$a" : "1$a";
                                          my $ob = -d $b ? "0$b" : "1$b";
                                          lc $oa cmp lc $ob }
                                grep{ $_ and $_ ne '.' and $_ ne '..' } readdir $dir;
        #say "fobj: ", join $/, @fobj;

        open my $fh, '>', $output_file;

        # create directory in pages only if
        # this directory contains a subdirectory

        # e.g. for path
        # /public/audio/foo/bar.mp3
        # /pages/audio/ needed, but not /pages/audio/foo

        my $target_dir = join '/',$target,@path,$current;
        mkdir $target_dir
                if ! -e $target_dir
                and grep { -d $_ } @fobj;

		scalar @fobj or say $fh "This directory is empty";
        for my $name( @fobj ){
                # link path (for URL)
                my $link_path = join_path('', @path, $current, $name);
                #$link_path =~ s/ /%20/g; # replace space with HTML entity
                #my $link = qq( <a href="). escape_uri($link_path).qq(">).escape_uri($name).qq(</a><br />\n);
		my $scrubbed_name = $name;
		$scrubbed_name =~ s/[^A-Za-z0-9\-_.!~*'() ]+//g;
		$scrubbed_name =~ s/ /_/g;
		my $dummy_name = "dummy";
                my $link = qq( <a href="). $link_path.qq(">).$name.qq(</a><br />\n);
                if( -d $name )  # directory
                {
                        opendir my($dirh), $name;
                        my @contents = grep{ $_ ne '.' and $_ ne '..' } readdir $dirh;
                        next unless @contents;
                        #say "name: $name";
                        #say "contents: @contents";
                        print $fh "(dir) $link";
                        #print $fh "test - ",++$i;
                        dir_list(
                                source => $source,
                                target => $target,
                                current => $name,
                                path => [@path, $current],
                        );

                        # recursive dir_list will have chdir'ed, so we need

                        chdir $full;

                }
                elsif( -f $name )       # plain file
                        {
                                my $size = -s $name;
                        #print $fh "test - ",++$i;
                                print $fh klength($size) . $link;
                        }


        }
        close $fh;
}
sub join_path {  join '/', @_ }

sub klength {
        my $n = shift;
        my $unit = " " x 2; # for bytes
        if($n > 1e9){ $n /= 1e9, $unit = 'GB'}
        elsif($n > 1e6){ $n /= 1e6, $unit = 'MB' }
        elsif($n > 1e3){ $n /= 1e3, $unit = 'KB' }
        sprintf("%.1f $unit", $n);
}
sub pad {
        my ($s, $len) = @_;
        sprintf("%*s", $len||10, $s);
}

sub salt {
	my $length = shift;
	my @chars=('a'..'z','A'..'Z','0'..'9','_');
	my $random_string;
	$random_string .= $chars[rand @chars] for (1..$length) ;
	$random_string;
}
sub generate_passwd {
	my $source = shift // 'pass';
	my $target = shift // 'passwd';
	open my $fh, '<', $source;
	open my $out, '>', $target;
	my $date = `date`;
	my ($seconds) = localtime(time);
	say srand(int($seconds**2/3));
	my $line; 
	while( $line = <$fh>){
	next if $line =~ /^#/;
	chomp $line; 
	my ($full, $email,$user,$pass, $allow) = split /\s*:\s*/, $line;
	next unless $allow =~ /yes/;
	my $salt = salt(8);
	my $crypt = apache_md5_crypt($pass, $salt);  # , $salt);
	say $out "$user $crypt $salt";
	}
}
sub generate_user_list {
	my @lines = read_file('pass');
	my $content = join $/,
 		map{ 
 			my $line = $_;
 			my (undef, $email, $user) = split /\s*:\s*/, $line;
 			"$email : $user"  
 		} @lines;
	write_file('users', $content);
}
sub find_user {
	my $user_or_email = shift;
	# we expect to be in Dancer app directory
	my @entries = read_file('users');
	my $body;
	my ($name, $email);
	map { chomp } @entries;
	foreach my $entry( @entries ){
		($name, $email) = split /\s*:\s*/, $entry;
		next if $user_or_email ne $name and $user_or_email ne $email;
		return { user => $name, email => $email};
	}
}
sub send_user_password {
	my $user_or_email = shift;
	my $subject = $website_name;
	# we expect to be in Dancer app directory
	my @entries = read_file('pass');
	my ($name, $email, $username, $password);
	map { chomp } @entries;
	foreach my $entry( @entries ){
		($name, $email, $username, $password) = split /\s*:\s*/, $entry;
		next if $user_or_email ne $username and $user_or_email ne $email;
		last 
	}
	my $body = mail_body($name, $username, $password);
	send_out_mail($email, $body);
}
sub remote_reminder { 
	my $email = shift; # could be email or username
	Email::Stuffer->from($sender)
				   ->to($email)
				   ->subject("CS3Reminder")
				   ->text_body($email)
				   ->send;
	
}
	
sub send_reminder {
	my $email = shift;
	my $result = find_passwd($email);
	if( $result )
	{ 
		send_out_mail($result->{email}, $result->{body});
	 	return 1
	}
}
sub send_out_mail {
	my( $email, $body) = @_; 
	Email::Stuffer->from($sender)
				   ->to($email)
				   ->subject("Reminder")
				   ->text_body($body)
				   ->send;
	
}

sub mail_body {
	my ($name, $username, $password) = @_;
	my ($first) = $name =~ /(.+?)\s*\S*$/;
	#say $first;
<<FORM;
Dear $first,

This automated mail contains your login credentials
for the $website_name training materials website.
Note that these credentials are for your exclusive use.
(Please don't share them.)

<a href=$website_url>$website_name</a>

Your username is $username
Your password is $password

Capitalization *is* signficant ("password" differs from
"PassWord")

Regards,

$webmaster_name
--
FORM
}

1
__END__


lib/Tutu.pm
package Tutu;
use Modern::Perl;
use Cwd;
use Data::Dumper::Concise;
use Dancer ':syntax';
use Dancer::Error;
use Dancer::Session;
use Crypt::PasswdMD5;
use Text::Markdown 'markdown';
use File::Basename 'fileparse';
use File::Slurp;
use File::Find::Rule;
use Tutu::Utils qw(index_directories find_user remote_reminder send_user_password);

use Cwd;
our $VERSION = 0.01;
our $ABSTRACT = "Tutu - a skimpy outfit for Dancer";
our ($root, $page_dir, $form_dir, $public, $access, $restricted,
	$passwd, %passwd, %salt, %access, %is_restricted, %render);
	
sub website_init {
	$root = shift // '.'; 
	$VERSION = '0.1';
	$page_dir = "$root/pages"; 
	$form_dir = "$root/forms"; 
	$public = "$root/public";
	$access = "$root/access";
	$passwd = "$root/passwd";
	$restricted = "$root/restricted_dirs";
	my @passwd = split "\n",read_file($passwd);
	while( my($user, $hash, $salt) = split " ", shift @passwd )
	{
		$passwd{$user} = $hash;
		$salt{$user} = $salt;
	}
	my @access = split "\n",read_file($access) if -e $access; 
	%access = map{ my($user,$access) = split(" ",$_,2); ($user,$access) } @access;
	%is_restricted = map{ $_, 1} split " ",read_file($restricted) if -e $restricted; 
	%render = ( html => sub{join '',@_},
				md  => sub{markdown(@_)}, 
				txt => sub{markdown(@_)});
}

hook 'before' => sub {
  	if ( 	
			(config->{require_login} and !session('user'))
			and request->path_info !~
			m{(login|bye|reminder-has-been-sent|remind|user-not-found)$}
	) {
  		# Pass the original path requested along to the handler:
  		session requested_path => request->path_info;
		redirect('/login');
  	}
  	elsif ( ! authorized( session('user'), request->path_info)
			and request->path_info ne '/forbidden'){
  		# Pass the original path requested along to the handler:
  		session requested_path => request->path_info;
		redirect('/forbidden');
  	}
};
sub authorized { 
	my ($user, $path) = @_;
	#say "user $user, path $path";
	#say "access: $access{$user}";
	#say Dumper \%access;
	my @parts = grep{ $_ } split '/', $path;
	# is this directory controlled?
	my ($restricted) = grep { $is_restricted{$_} } @parts;
	#say "restricted $restricted";
	return 1 unless $restricted;
	#grep { $restricted eq $_ } split " ",$access{$user}
	#die "path: $path, $access{$user}" if session('user') eq 'joel';
	#die "$access{$user}" if session('user') eq 'joel';
	grep{ $path =~ /\b$_\b/ } split " ",$access{$user}
}

post '/login'    => sub {
	no warnings 'uninitialized';
	if(	exists $passwd{params->{user}} 
			and apache_md5_crypt(params->{pass},$salt{params->{user}}) 
				eq $passwd{params->{user}})
	{
		session user => params->{user};
	my $path = 	'/welcome';
#   	$path = session->{requested_path}
#   		if session->{requested_path} =~ /\w+/ 
# 			and session->{requested_path} !~ /login/;
		session requested_path => undef;
		session failed_login => undef;
		redirect $path;
	} else {
		session failed_login => 1;
		redirect '/login'
	}
};
get  '/login'    => sub { 
	session ident => undef; # if leftover from password reminder
	session user_not_found => undef;
	preprocess('login',"$page_dir/login.md",$render{'md'})};
get  '/logout'    => sub { 
			session->destroy;
			redirect '/bye' };

get '/' => sub { redirect '/welcome' };

get '/forbidden' => sub { forbidden(session->{requested_path}) };

## The following two routes are a hack. The get '/**' route should
## render the /upload route (if user is allowed) and *never* return
## a page for /upload.html

get '/**' => sub {
	# there is nothing in the /public directory
	# so we serve all requests

	no warnings 'uninitialized';
	my ($matches) = splat;
	my $filename = pop @$matches;

### This code replaces the default serving from /public 
# 	by placing the /public tree under /public/floor
# 
# 	my $search_dir = join '/', $public, 'floor', @$matches;
# 	# first look for a matching file to return
# 
# 	my ($filepath) = File::Find::Rule
#                 ->file()
#                 ->name($filename)
#                 ->in($search_dir);
# 	
# 	my @parts = split '/',$filepath;
# 	shift @parts if $parts[0] eq 'public';
# 	#die Dumper \@parts; # floor images chicago400.jpg
# 	if ($filepath){
# 		my ($ext) = $filename =~ /\.(\w+)$/;
# 		my %force_download = map{$_ => 1 } qw(mpeg mp3 m4v mp4 mpg dvd iso zip);
# 		my @args;
# 		headers 'Content-disposition', qq(attachment; filename="$filename")
# 			if $force_download{$ext};	
# 		content_type 'application/octet-stream';
# 		
# 		# /floors/image/chicago400.jpg
# 		send_file(join('/', undef,@parts), @args);
# 	}
	if (0){}
	else # maybe we have an .html .txt or .md file to serve ?
	{ 
				my $dir = join '/',$page_dir, @$matches;
				my ($file) = File::Find::Rule->file()
                                     ->name( "$filename*" )
                                     ->in($dir);
				if ( ! $file ) { err_not_found($file) }
				else { 
					my ($name, $path, $ext) = fileparse($file,qw(html md));
					preprocess($name, $file, $render{$ext});
				}
	}
				
};
post '/remind'    => sub {
	my $ident = params->{ident};
 	my $result = find_user($ident) ;
	session user_not_found => undef;
 	if ($result)
	{ 
		if ( -f 'pass' ) # we have the password file on hand
			{ send_user_password($result->{user}) }
		else { remote_reminder($result->{user}) }
		session ident => $ident;
		redirect '/reminder-has-been-sent';
	}
 	else 
	{ 
 		session user_not_found => $ident;
	  	redirect '/remind'
 	}
};
post '/upload-file'    => sub {
	
	my ($name, $section, $description) ;
	{ 
		no warnings 'uninitialized';
		($name, $section, $description) 
			= (params->{filename}, params->{category}, params->{description});
	}
	my $file = request->upload('filename');
	defined $file or redirect '/upload-failed';
	my $destination = "$root/uploads";
	$destination .= $section if $section;
	$destination .= "/$name";
	$file->copy_to($destination);
	session uploaded_file_path => $destination;
	session uploaded_file_name => $name;
	redirect '/upload-succeeded';
};
sub err_not_found { 
	my $path = shift;
	final_output("Error 404: file not found", 
qq(

The following file or page you requested was not found:
<p>
<h2>$path</h2>
<p>
Please contact the webmaster if you think
this message was received in error.
)
)
}

sub forbidden { 
	my $path = shift;
	final_output("Error 404: forbidden", 
qq(

Your account lacks permission to access the following file or page:
<p>
<h2>$path</h2>
<p>
Please contact the webmaster if you think
this message was received in error.
)
)
}

sub preprocess {
		my ($topic, $path, $preprocess) = @_;
		# $topic:      page name
		# $preprocess: coderef to return content
		$preprocess //= sub { @_ }; # return content unchanged
		if (-r $path){  
			my $text = read_file($path);
			my $body = $preprocess->($text);
			final_output($topic, $body);
		}
		else { err_not_found($path) }
}

sub final_output {
		my ($topic, $body) = @_;
		template 'default.tt', 
				 {body  => $body,
				  topic => $topic,
				 }, 
				 {layout => undef };

}
true;
__END__

=pod

=head1 NAME

Tutu - a skimpy outfit for Dancer 

=head1 SYNOPSIS


	cd $TUTU_BUILD_DIRECTORY

	$EDITOR MY_WEBSITE_INIT

	tutu-configure MY_WEBSITE_INIT

	# move $WEBSITE_ROOT and $WEBSITE_DOMAIN directories 
	# to appropriate places if necessary
	
	cd $WEBSITE_ROOT

	./dance.psgi # point your browser at http://localhost:3000


=head1 DESCRIPTION

Outfitted with Tutu, Dancer performs page
rendering, access control and file up/downloading. Tutu
also includes utilities for:

=over 8

=item tutu-genapp : create and configure a new webapp 

=item tutu-index : generate file directory listings

=item tutu-pass : generate passwd file

=item tutu-sync : synchronize the website with local development environment

=back

=head1 LICENSE

This module is free software and is published under the same
terms as Perl itself.

=cut
